restapi/src/main/java/io/stargate/web/docsapi/service/JsonSchemaHandler.java

```diff
@@ -0,0 +1,138 @@
+package io.stargate.web.docsapi.service;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.github.fge.jsonschema.core.exceptions.ProcessingException;
+import com.github.fge.jsonschema.core.report.ProcessingReport;
+import com.github.fge.jsonschema.main.JsonSchemaFactory;
+import io.stargate.db.schema.Schema;
+import io.stargate.web.docsapi.dao.DocumentDB;
+import io.stargate.web.docsapi.exception.ErrorCode;
+import io.stargate.web.docsapi.exception.ErrorCodeRuntimeException;
+import io.stargate.web.docsapi.models.JsonSchemaResponse;
+import io.stargate.web.docsapi.service.util.ImmutableKeyspaceAndTable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.inject.Inject;
+
+/** Service that handles attaching and fetching JSON schemas to/from Documents API collections. */
+public class JsonSchemaHandler {
+  private ObjectMapper mapper;
+  private final JsonSchemaFactory schemaFactory = JsonSchemaFactory.byDefault();
+  private final ConcurrentHashMap<ImmutableKeyspaceAndTable, JsonNode> schemasPerCollection =
+      new ConcurrentHashMap<>();
+  private Schema lastCheckedSchema;
+
+  @Inject
+  public JsonSchemaHandler(ObjectMapper mapper) {
+    if (mapper == null) {
+      throw new IllegalStateException("JsonSchemaHandler requires a non-null ObjectMapper");
+    }
+    this.mapper = mapper;
+  }
+
+  private void clearCacheOnSchemaChange(DocumentDB db) {
+    if (!db.schema().equals(lastCheckedSchema)) {
+      schemasPerCollection.clear();
+      this.lastCheckedSchema = db.schema();
+    }
+  }
+
+  private JsonSchemaResponse reportToResponse(JsonNode schema, ProcessingReport report) {
+    JsonSchemaResponse resp = new JsonSchemaResponse(schema);
+    report.forEach(
+        msg -> {
+          resp.addMessage(msg.getLogLevel(), msg.getMessage());
+        });
+    return resp;
+  }
+
+  public JsonSchemaResponse attachSchemaToCollection(
+      DocumentDB db, String namespace, String collection, JsonNode schema) {
+    ProcessingReport report = schemaFactory.getSyntaxValidator().validateSchema(schema);
+    JsonSchemaResponse resp = reportToResponse(schema, report);
+    if (report.isSuccess()) {
+      ObjectNode wrappedSchema = mapper.createObjectNode();
+      wrappedSchema.set("schema", schema);
+      writeSchemaToCollection(db, namespace, collection, wrappedSchema.toString());
+      ImmutableKeyspaceAndTable info =
+          ImmutableKeyspaceAndTable.builder().keyspace(namespace).table(collection).build();
+      schemasPerCollection.put(info, schema);
```

Comment:
In the face of potentially concurrent "attach" operations, I believe it would be more correct to do `.remove(info)` here and let `getCachedJsonSchema()` fetch the latest schema on request.